### PR TITLE
Feature/OpenSea: Create OpenSea package

### PR DESCRIPTION
**Pull request by dieudonneAwa**

### CODE REVIEW

File: packages/OpenSea/index.js [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        </span>
+        <div style="margin-top: 0.6rem;" class="h-full w-full">
+          <div style="padding: 1rem 0; flex: 1;" class="flex flex-col justify-start items-start">
+            <a href="${
+              firstAsset ? firstAsset.permalink || firstAsset.external_link : ""
+            }">
+              <button class="btn--primary cursor-pointer flex items-center rounded" style="border: 0; padding: 0.8rem 2.5rem; margin-bottom: 1rem;">
+                VIEW THIS ITEM
+                <span style="margin-left: 0.5rem;" class="material-icons">
+                  keyboard_arrow_right
```
Comment (treygrainger) [inline packages/OpenSea/index.js:103]: This "keyboard_arrow_right" text is rendering directly since the styles weren't added to the package and were added to the demo page instead. Please ensure any needed styles get moved into this package (not an external reference) so it's self-contained. Thanks!

image

File: packages/OpenSea/index.js [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+                </span>
+              </button>
+            </a>
+          </div>
+        </div>
+      </div>
+    </div>
+
+    <div style="max-width: 500px;" class="mx-2 rounded flex flex-column container-fluid bg-cultured">
```
Comment (treygrainger) [inline packages/OpenSea/index.js:112]: I don't think we're going to be able to really display more than one of these unless the other ones are way smaller and organized. I'd suggest maybe only showing the first one, and then if there are more than one have a "See more >" link of some sort. This is what is currently looks like, for reference:

image

File: packages/OpenSea/index.js [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        }</span>
+        <span class="flex items-center text-grey-web" style="margin-top: 1rem;">
+          <span style="height: 25px; width: 25px; overflow: hidden; border-radius: 50%; margin-right: 0.5rem;" class="flex justify-center items-center">
+            <img src="${
+              firstAsset && firstAsset.owner
+                ? firstAsset.owner.profile_img_url
+                : ""
+            }" width="40px" />
+          </span>
+          Owned by &nbsp; <a class="text-bluetiful" href="${
```
Comment (treygrainger) [inline packages/OpenSea/index.js:85]: The "owned by" is always showing up as "Address" for me, which looks a little strange since we know the owning address and are linking to it. I think we should show the actual address instead, or possibly a shortened version to save space (i.e. 0xXXXX...XXXX)

File: packages/OpenSea/index.js [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+                .join("")
+            : ""
+        }
+      </div>
+    </div>
+  </div>`;
+}
+
+async function trigger(query) {
+  const assets = await loadAssets(query);
```
Comment (treygrainger) [inline packages/OpenSea/index.js:149]: This is running the API calls to OpenSea for every single search request, which we absolutely cannot do. The purpose of the trigger is to have a small, light-weight check to verify whether we should invoke the package at all. In this case, the trigger is currently hitting the opensea API twice for every query (i.e. "shoes") just to check if it's an address with an NFT. Then it later hits the opensea API again during processing another 2 times.

The trigger function needs to be replaces by something that validates whether the query could trigger the package. In this case, you probably want a light-weight Regex that matches all valid contract addresses across the various blockchains that could hold NFTs.

We definitely can't be hitting APIs in the trigger, though.

File: packages/OpenSea/index.js [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+  if (
+    assets &&
+    assets.length &&
+    assets.some((asset) => {
+      return (
+        asset &&
+        ((asset.owner && asset.owner.address === query) ||
```
Comment (treygrainger) [inline packages/OpenSea/index.js:156]: Same question as above as to whether the payout address and creator.address are redundant given the actual API checks only checking the other two.

File: packages/OpenSea/index.js [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+          (asset.creator && asset.creator.address === query))
+      );
+    })
+  ) {
+    return true;
+  }
+  return false;
+}
+
+module.exports = { OpenSea, trigger };
```
Comment (treygrainger) [inline packages/OpenSea/index.js:168]: will be "openSea" after you rename function

File: packages/OpenSea/package.json [hunk 7]
```diff
@@ -0,0 +1,11 @@
+{
+  "name": "opensea",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.ts",
```
Comment (treygrainger) [inline packages/OpenSea/package.json:5]: Should be index.js and not index.ts

File: server/index.js [hunk 8]
```diff
@@ -12,16 +12,20 @@ const readFileAsync = promisify(fs.readFile);
 const app = express();
 const PORT = 4000;
 
-app.get('/', async (req, res) => {
+app.get("/", async (req, res) => {
   const commonCss = await readFileAsync(`${__dirname}/../common.css`);
   const packageKey = req.query.packageKey;
   const query = req.query.query;
+  console.log(packageObject);
```
Comment (treygrainger) [inline server/index.js:19]: Remove this. Looks like it was added during debugging.

File: server/index.js [hunk 9]
```diff
@@ -12,16 +12,20 @@ const readFileAsync = promisify(fs.readFile);
...
   const trigger = await packageObject[packageKey].trigger(query);
   if (packageKey in packageObject && trigger) {
     const result = await packageObject[packageKey][packageKey](query);
     res.send(`
       <head>
         <meta name="viewport" content="width=device-width,initial-scale=1,shrink-to-fit=no">
+        <link href="https://fonts.googleapis.com/css2?family=Material+Icons"
```
Comment (treygrainger) [inline server/index.js:26]: Can't import styles like this here for a a few reasons:
1) This is just a demo page so you can see what a package will render like. Any styles added to the demo page and not the package itself won't apply to the package, making it just look good on the demo page and not when the actual package is rendered in production.
2) All packages need to have self-contained styles. Applying global styles can impact the rendering of all packages.
3) We can't link the third-party files like this anyway, as it allows third parties to track users. In this case, Google and jsdelivr.net could both track our users if we link external files like this in the package.

So TLDR; we need to keep this index.js page free from any styles that would make it look different than production, need to ensure each package's styles are self-contained within the package, and need to ensure we're not linking to any files that would require a user's web browser to hit a third-party website.

**Review by treygrainger — CHANGES_REQUESTED** [review]
Hi @dieudonneAwa, thanks for putting this package together! I've added comments throughout the PR with requested changes. If you have any questions, feel free to ask!

All the best,

-Trey

File: packages/OpenSea/index.js [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+  const assets = await loadAssets(query);
+
+  const firstAsset =
+    assets &&
+    assets.find((asset) => {
+      return (
+        asset &&
+        ((asset.owner && asset.owner.address === query) ||
+          (asset.asset_contract && asset.asset_contract.address === query) ||
+          (asset.collection && asset.collection.payout_address === query) ||
```
[the remark this answers is not in the document]
Comment (dieudonneAwa) [reply packages/OpenSea/index.js:31]: Yes @treygrainger 
You're right about the checks on collection.payout_address and asset.creator addresses, they are not necessary. I will remove them at once.

**Comment by jejopl** [comment]
@dieudonneAwa I've moved this PR to #88
Let me know if everything is okay and I will merge it

**Comment by dieudonneAwa** [comment]
> @dieudonneAwa I've moved this PR to #88 Let me know if everything is okay and I will merge it

All okay I think.
